include/cetl/rtti.hpp

```diff
@@ -0,0 +1,316 @@
+/// This won't work in the presence of multiple inheritance because it makes implicit conversion
+/// to the base class ambiguous.
+template <typename T>
+constexpr bool has_polymorphic_type_id = decltype(detail::has_polymorphic_type_id_impl<std::decay_t<T>>(0))::value;
+
+/// True iff \c T has a public method \c _cast_().
+template <typename T>
+constexpr bool is_rtti_convertible = decltype(detail::has_cast_impl<std::decay_t<T>>(0))::value;
+
+/// A helper that converts \ref cetl::type_id_type to \ref cetl::type_id.
+template <typename TypeIDType>
+constexpr type_id type_id_type_value = detail::type_id_type_value_impl(TypeIDType{});
+
+/// An alternative implementation of simple runtime type information (RTTI) capability designed for high-integrity
+/// real-time systems, where the use of the standard C++ RTTI is discouraged.
+///
+/// Unlike the standard C++ RTTI, this implementation requires the user to manually specify the type ID per type
+/// in the form of a 16-byte UUID (GUID). The capabilities include querying runtime type information in constant time
+/// and performing safe dynamic type down-conversion (and up-conversion, similar to \c dynamic_cast) in constant time,
+/// including the case of multiple inheritance.
+///
+/// The limitations are that a type has to opt into this RTTI capability explicitly (it doesn't work for all types)
+/// and that CV-qualifiers and references are not considered in the type comparison;
+/// that is, <tt>T</tt> and <tt>const T&</tt> are considered the same type.
+///
+/// In order to support this RTTI capability, a type must at least define a public method
+/// <tt>static constexpr cetl::type_id _get_static_type_id_() noexcept</tt> that
+/// returns the type ID (i.e., the UUID) of the type it is defined on.
+/// Types that provide said method satisfy \ref cetl::has_static_type_id.
+/// If the type is polymorphic, it should publicly inherit from \ref cetl::rtti_helper, which will provide the
+/// necessary implementation of the RTTI-related methods along with the aforementioned
+/// \c _get_static_type_id_() method.
+///
+/// RTTI-related members are named with surrounding underscores to avoid name clashes with user-defined members,
+/// including the members of DSDL types (where identifiers surrounded with underscores are reserved).
+/// The user code should not invoke such underscored methods directly but instead use the free functions defined here.
+class rtti
+{
+public:
+    /// Implementations are adivsed to use \ref cetl::rtti_helper instead of implementing this manually.
+    /// If manual implementation is preferred, then the method body should simply forward the result of this
+    /// type's static \c _get_static_type_id_ method:
+    ///
+    /// @code
+    /// return _get_static_type_id_();
+    /// @endcode
+    ///
+    /// This method should not be invoked directly; instead, use \ref cetl::get_type_id.
+    CETL_NODISCARD virtual type_id _get_polymorphic_type_id_() const noexcept = 0;
```

Comment:
Which ID is returned though? Polymorphic objects can be multiple types. That's the whole point. By asking "are you this type" you don't have to expose a list of types to the user which is what you'll have to do if you are going to ask for types from the object.

By moving away from the "queryType" method you remove the ability for implementations to be returned directly instead of using delegated interfaces. This might be okay. It does simplify the type system even if it means certain optimizations, normally available to COM objects, are not supported and it means we don't add anything that rtti doesn't provide.

Reply:
>Which ID is returned though? Polymorphic objects can be multiple types.

The ID of the most concrete type, aka runtime type, is returned, which is the behavior of `dynamic_cast`. I can spell it out in the docs to avoid misinterpretations.

>By moving away from the "queryType" method you remove the ability for implementations to be returned directly instead of using delegated interfaces.

I attempted to deliver the functionality offered by queryType in this design via the `_cast_` method, and its user-facing facade `rtti_cast`. Do you think this design does not offer equivalent capabilities to the user?